perf/uncore_libpfm.go

```diff
@@ -0,0 +1,369 @@
+
+	for _, group := range c.events.MemoryController {
+		customEvent, ok := c.eventToCustomEvent[group[0]]
+		var err error
+		if ok {
+			err = c.setupRawNonGroupedUncoreMemoryController(customEvent)
+		} else {
+			err = c.setupNonGroupedUncoreMemoryController(string(group[0]))
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func parseUncoreEvents(events Events) (UncoreEvents, map[Event]*CustomEvent) {
+	eventToCustomEvent := map[Event]*CustomEvent{}
+	for _, uncoreEvent := range events.UncoreEvents.MemoryController {
+		for _, customEvent := range events.CustomEvents {
+			if uncoreEvent[0] == customEvent.Name {
+				eventToCustomEvent[customEvent.Name] = &customEvent
+			}
+		}
+	}
+	return events.UncoreEvents, eventToCustomEvent
+}
+
+func (c *uncoreCollector) Destroy() {
+	c.cpuFilesLock.Lock()
+	defer c.cpuFilesLock.Unlock()
+
+	for name, pmus := range c.cpuFiles {
+		for pmu, cpus := range pmus {
+			for cpu, file := range cpus {
+				klog.V(5).Infof("Closing uncore perf_event file descriptor for event %q, type %q and CPU %d", name, pmu, cpu)
+				err := file.Close()
+				if err != nil {
+					klog.Warningf("Unable to close perf_event file descriptor for event %q, type %q and CPU %d", name, pmu, cpu)
+				}
+			}
+			delete(pmus, pmu)
+		}
+		delete(c.cpuFiles, name)
+	}
+}
+
+func (c *uncoreCollector) GetStats() ([]info.PerfStat, error) {
+	stats := []info.PerfStat{}
+
+	klog.V(5).Info("Attempting to update uncore perf_event stats")
+
+	for name, pmus := range c.cpuFiles {
+		for pmu, cpus := range pmus {
+			for cpu, file := range cpus {
+				buf := make([]byte, 32)
+				_, err := file.Read(buf)
+				if err != nil {
+					klog.Warningf("Unable to read from uncore perf_event file (event: %q, CPU: %d)", name, cpu)
+					continue
+				}
+				perfData := &ReadFormat{}
+				reader := bytes.NewReader(buf)
+				err = binary.Read(reader, binary.LittleEndian, perfData)
+				if err != nil {
+					klog.Warningf("Unable to decode from binary format read from uncore perf_event file (event: %q, CPU: %d)", name, cpu)
+					continue
+				}
+				klog.V(5).Infof("Read metric for event %q for cpu %d and pmu %d : %d", name, cpu, pmu, perfData.Value)
+				scalingRatio := 1.0
+				if perfData.TimeEnabled != 0 {
+					scalingRatio = float64(perfData.TimeRunning) / float64(perfData.TimeEnabled)
+				}
+				stat := info.PerfStat{
+					Value:        uint64(float64(perfData.Value) / scalingRatio),
+					Name:         name,
+					ScalingRatio: scalingRatio,
+					Cpu:          cpu,
+					Type:         pmu,
+				}
+				stats = append(stats, stat)
+			}
+		}
+	}
+
+	return stats, nil
+}
+
+func (c *uncoreCollector) setupRawNonGroupedUncoreMemoryController(event *CustomEvent) error {
+	klog.V(5).Infof("Setting up non-grouped raw perf uncore memory controller event %#v", event)
+
+	pmus, err := getIMCPMUs()
+	if err != nil {
+		return err
+	}
+
+	if event.Type == 0 {
+		// PMU isn't set. Register event for all memory PMUs.
```

Comment:
It must be clearly stated in the documentation.

Reply:
It's looking good now?

Reply:
according kernel [perf_type_id](https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/perf_event.h#L29) "0" value means hardware counter 

it is a little confusing why "0" here has other meaning, could you explain this?